nova_vm/src/engine/bytecode/executable/function_declaration_instantiation.rs

```diff
@@ -0,0 +1,334 @@
+            // ii. NOTE: In strict mode code early errors prevent attempting to assign to this binding, so its mutability is not observable.
+            ctx.exe.add_instruction_with_identifier(
+                Instruction::CreateImmutableBinding,
+                BUILTIN_STRING_MEMORY.arguments,
+            );
+        } else {
+            // d. Else,
+            //   i. Perform ! env.CreateMutableBinding("arguments", false).
+            ctx.exe.add_instruction_with_identifier(
+                Instruction::CreateMutableBinding,
+                BUILTIN_STRING_MEMORY.arguments,
+            );
+        }
+
+        // e. Perform ! env.InitializeBinding("arguments", ao).
+        ctx.exe.add_instruction_with_identifier(
+            Instruction::ResolveBinding,
+            BUILTIN_STRING_MEMORY.arguments,
+        );
+        ctx.exe
+            .add_instruction(Instruction::InitializeReferencedBinding);
+
+        // f. Let parameterBindings be the list-concatenation of parameterNames and « "arguments" ».
+        // NOTE: reusing `parameter_names` for `parameterBindings`.
+        parameter_names.insert("arguments".into());
+    }
+
+    // 24. Let iteratorRecord be CreateListIteratorRecord(argumentsList).
+    // 25. If hasDuplicates is true, then
+    //   a. Perform ? IteratorBindingInitialization of formals with arguments iteratorRecord and undefined.
+    // 26. Else,
+    //   a. Perform ? IteratorBindingInitialization of formals with arguments iteratorRecord and env.
+    if !formals.has_parameter() {
+        ctx.exe.add_instruction(Instruction::IteratorClose)
+    } else if has_parameter_expressions {
+        complex_array_pattern(
+            ctx,
+            formals.items.iter().map(|param| Some(&param.pattern)),
+            formals.rest.as_deref(),
+            !has_duplicates,
+        );
+    } else {
+        simple_array_pattern(
+            ctx,
+            formals.items.iter().map(|param| Some(&param.pattern)),
+            formals.rest.as_deref(),
+            formals.items.len(),
+            !has_duplicates,
+        );
+    }
+
+    // 27. If hasParameterExpressions is false, then
+    if !has_parameter_expressions {
+        // a. NOTE: Only a single Environment Record is needed for the parameters and top-level vars.
+        // b. Let instantiatedVarNames be a copy of the List parameterBindings.
+        let mut instantiated_var_names = AHashSet::new();
+        // c. For each element n of varNames, do
+        for n in function_body_var_declared_names(body) {
+            // i. If instantiatedVarNames does not contain n, then
+            if instantiated_var_names.contains(&n) || parameter_names.contains(&n) {
+                continue;
+            }
+            // 1. Append n to instantiatedVarNames.
+            let n_string = String::from_str(&mut ctx.agent, &n);
+            instantiated_var_names.insert(n);
+            // 2. Perform ! env.CreateMutableBinding(n, false).
+            ctx.exe
```

Comment:
**question:** Isn't this the `CreateAndInitializeVariableBinding` instruction you added?

Reply:
If `!has_parameter_expressions`, then the variables are initialized in `env`, which is the *lexical* environment; otherwise, they are initialized in `varEnv`, which is the *variable* environment (and might also be set to the lexical environment or not afterwards depending on `strict`).

I added that instruction in order to do the `varEnv` initialization if `has_parameter_expressions`, but then I decided to add `InitializeVariableEnvironment` instead, and I forgot to remove `CreateAndInitializeVariableBinding`.

Reply:
I just removed the `CreateAndInitializeVariableBinding` instruction.